go/store/datas/pull/pull_chunk_fetcher.go

```diff
@@ -0,0 +1,174 @@
+
+	"github.com/dolthub/dolt/go/store/hash"
+	"github.com/dolthub/dolt/go/store/nbs"
+)
+
+type GetManyer interface {
+	GetManyCompressed(ctx context.Context, hashes hash.HashSet, found func(context.Context, nbs.CompressedChunk)) error
+}
+
+// A ChunkFetcher is a batching, stateful, potentially concurrent interface to
+// fetch lots of chunks from a ChunkStore. A caller is expected to call
+// `Get()` and `Recv()` concurrently. Unless there is an error, for every
+// single Hash passed to Get, a corresponding Recv() call will deliver the
+// contents of the chunk. When a caller is done with a ChunkFetcher, they
+// should call |CloseSend()|. After CloseSend, all requested hashes have been
+// delivered through Recv(), Recv() will return `io.EOF`.
+//
+// A ChunkFetcher should be Closed() when it is no longer needed. In non-error
+// cases, this will typically be after Recv() has delivererd io.EOF. If Close
+// is called before Recv() delivers io.EOF, there is no guarantee that all
+// requested chunks will be delivered through Recv().
+//
+// In contrast to interfaces to like GetManyCompressed on ChunkStore, if the
+// chunk is not in the underlying database, then Recv() will return an
+// nbs.CompressedChunk with its Hash set, but with empty contents.
+//
+// Other than an io.EOF from Recv(), any |error| returned from any method
+// indicates an underlying problem wtih fetching requested chunks from the
+// ChunkStore. A ChunkFetcher is single use and cannot be used effectively
+// after an error is returned.
+type ChunkFetcher interface {
+	Get(ctx context.Context, hashes hash.HashSet) error
+
+	CloseSend() error
+
+	Recv(context.Context) (nbs.CompressedChunk, error)
+
+	Close() error
+}
+
+// A PullChunkFetcher is a simple implementation of |ChunkFetcher| based on
+// calling GetManyCompressed.
+//
+// It only has one outstanding GetManyCompressed call at a time.
+type PullChunkFetcher struct {
+	ctx context.Context
+	eg  *errgroup.Group
+
+	getter GetManyer
+
+	batchCh chan hash.HashSet
+	doneCh  chan struct{}
+	resCh   chan nbs.CompressedChunk
+}
+
+func NewPullChunkFetcher(ctx context.Context, getter GetManyer) *PullChunkFetcher {
+	eg, ctx := errgroup.WithContext(ctx)
+	ret := &PullChunkFetcher{
+		ctx:     ctx,
+		eg:      eg,
+		getter:  getter,
+		batchCh: make(chan hash.HashSet),
+		doneCh:  make(chan struct{}),
+		resCh:   make(chan nbs.CompressedChunk),
+	}
+	ret.eg.Go(func() error {
+		return ret.fetcherThread(func() {
+			close(ret.resCh)
+		})
+	})
+	return ret
+}
+
+func (f *PullChunkFetcher) fetcherThread(finalize func()) error {
+	for {
+		select {
+		case batch, ok := <-f.batchCh:
+			if !ok {
+				finalize()
+				return nil
+			}
+
+			var mu sync.Mutex
+			missing := batch.Copy()
+
+			// Blocking get, no concurrency, only one fetcher.
+			err := f.getter.GetManyCompressed(f.ctx, batch, func(ctx context.Context, chk nbs.CompressedChunk) {
+				mu.Lock()
+				missing.Remove(chk.H)
+				mu.Unlock()
+				select {
```

Comment:
This select block is a little odd to me. Why aren't you just sticking chk into f.resCh and calling it a day? These are unlimited channels, so they aren't going to block, right? Does this avoid getting a panic when the channel has been closed prematurely?

Reply:
The channels are blocking and unbuffered...it's a synchronous exchange. So here we don't have a way to bail early but the reader might have already shut down, so we need to get through the batch of responses by ignoring them.